### PR TITLE
Implement a WtxId struct, and use it in Zebra's external network protocol

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

In order to handle wide transaction IDs from the network, Zebra needs to:
-  split them into `id` and `auth_data` fields
- use them when serializing and deserializing inventory hashes (in `inv`, `getdata`, and `notfound` messages)

### Specifications

https://zips.z.cash/zip-0239#specification

## Solution

Changes
- Add a `WtxId` type for wide transaction IDs
- Add conversions between transaction IDs and bytes
- Use the `WtxId` type in external network protocol messages 

Bug Fixes
- Make the `AuthDigest` display order match transaction IDs

Tests
- Move a `transaction::Hash` test to tests module
- Add display order tests for transaction IDs and auth digests
- Add round-trip byte and struct tests for transaction IDs and auth digests

This is part of ticket #2449, but it does not close that ticket.

## Review

Anyone can review this PR.
It's blocking all of the mempool work that needs wide transaction IDs.

It's based on PR #2547, so it might need rebasing after that PR merges to `main`.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Create an `UnminedTxId` enum, to abstract over unmined v4 and v5 transactions in Zebra's internal network protocol
  - `UnminedTxId` is also blocking all of the mempool work that needs wide transaction IDs

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2618)

### CODE REVIEW

File: zebra-network/src/protocol/external/inv.rs [hunk 1]
```diff
@@ -45,8 +46,21 @@ pub enum InventoryHash {
     /// [auth_digest]: https://zips.z.cash/zip-0244#authorizing-data-commitment
     /// [zip239]: https://zips.z.cash/zip-0239
     /// [bip339]: https://github.com/bitcoin/bips/blob/master/bip-0339.mediawiki
-    // TODO: Actually handle this variant once the mempool is implemented
-    Wtx([u8; 64]),
+    // TODO: Actually handle this variant once the mempool is implemented (#2449)
```
Comment (teor2345) [inline zebra-network/src/protocol/external/inv.rs:49]: We haven't finished this TODO, but we won't need to change this code again as part of #2449.

```suggestion
```

File: zebra-network/src/protocol/external/inv.rs [hunk 2]
```diff
@@ -63,39 +77,43 @@ impl From<block::Hash> for InventoryHash {
...
+}
+
 impl ZcashSerialize for InventoryHash {
     fn zcash_serialize<W: Write>(&self, mut writer: W) -> Result<(), std::io::Error> {
-        let (code, bytes): (_, &[u8]) = match self {
-            InventoryHash::Error => (0, &[0; 32]),
-            InventoryHash::Tx(hash) => (1, &hash.0),
-            InventoryHash::Block(hash) => (2, &hash.0),
-            InventoryHash::FilteredBlock(hash) => (3, &hash.0),
-            InventoryHash::Wtx(bytes) => (5, bytes),
```
Comment (teor2345) [inline zebra-network/src/protocol/external/inv.rs:73]: Once we stop storing `WtxId`s as a byte array, it gets much harder to return a slice here.
(We'd have to create an extra binding to store the temporary clone.)

File: zebra-chain/src/transaction/auth_digest.rs [hunk 3]
```diff
@@ -1,20 +1,99 @@
...
+};
 
 use super::Transaction;
 
 /// An authorizing data commitment hash as specified in [ZIP-244].
 ///
-/// [ZIP-244]: https://zips.z.cash/zip-0244..
-#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+/// Note: Zebra displays transaction and block hashes in big-endian byte-order,
+/// following the u256 convention set by Bitcoin and zcashd.
```
Comment (dconnolly) [inline zebra-chain/src/transaction/auth_digest.rs:18]: 👍

File: zebra-chain/src/transaction/auth_digest.rs [hunk 4]
```diff
@@ -1,20 +1,99 @@
...
+    fn from(auth_digest: &AuthDigest) -> Self {
+        (*auth_digest).into()
+    }
+}
+
+impl fmt::Display for AuthDigest {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut reversed_bytes = self.0;
+        reversed_bytes.reverse();
+        f.write_str(&hex::encode(&reversed_bytes))
```
Comment (dconnolly) [inline zebra-chain/src/transaction/auth_digest.rs:70]: 👍

**Review by conradoplg — APPROVED** [review]
Looks good, I liked the new `code()`

File: zebra-network/src/protocol/external/inv.rs [hunk 5]
```diff
@@ -45,8 +46,21 @@ pub enum InventoryHash {
     /// [auth_digest]: https://zips.z.cash/zip-0244#authorizing-data-commitment
     /// [zip239]: https://zips.z.cash/zip-0239
     /// [bip339]: https://github.com/bitcoin/bips/blob/master/bip-0339.mediawiki
-    // TODO: Actually handle this variant once the mempool is implemented
-    Wtx([u8; 64]),
+    // TODO: Actually handle this variant once the mempool is implemented (#2449)
+    Wtx(transaction::WtxId),
```
Comment (dconnolly) [inline zebra-network/src/protocol/external/inv.rs:50]: 👍

**Review by dconnolly — COMMENTED** [review]
👍

**Pull request merged**